### PR TITLE
Bst in memory

### PR DESCRIPTION
**Pull request by shreyasseshadri**

### CODE REVIEW

File: utils/bst/bst.cpp [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+
+        // If `node` is NULL that means we will be inserting below `prev` node,
+        // and hence we will not be unlocking it here, and will unlock once after we insert new node
+    }
+
+    Node *new_node = init_node(key, value);
+
+    if (!prev)
+    {
+        root = new_node;
```
Comment (srinskit) [inline utils/bst/bst.cpp:103]: You need to sync this. General rule, consider sync for every assignment.

File: utils/bst/bst.cpp [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+Node *Bst::upsert_node(std::string key, std::string value)
+{
+    Node *node = root;
+    Node *prev = NULL;
+
+    while (node)
+    {
+        // Lock from here, multiple threads should not insert nodes at the same time,
+        // Might lead to overriding paths
+        pthread_mutex_lock(&node->lock);
```
Comment (srinskit) [inline utils/bst/bst.cpp:72]: You need to sync here. Lock the BST, access root, lock root, release BST. Else, there is nothing protecting node from being de-allocated between 65 and 72. When you access something you need to lock it's parent first, as you have done for the traversal.

File: utils/bst/bst.cpp [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+            return node;
+        }
+        else
+        {
+            node = node->right;
+        }
+
+        if (node)
+            // After we moved `node` ptr, we can unlock nodes which we will not insert below
+            pthread_mutex_unlock(&prev->lock);
```
Comment (srinskit) [inline utils/bst/bst.cpp:93]: There are gaps in your locks.
```
node.lock()
child = node.left
node.unlock()

<another thread delete your child>

child.lock()
```

File: utils/bst/bst.cpp [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+
+    return node;
+}
+
+Bst::Bst()
+{
+    root = NULL;
+}
+
+Bst::~Bst()
```
Comment (srinskit) [inline utils/bst/bst.cpp:21]: You need to lock every node before delete

File: utils/bst/bst.cpp (hunk 2 above)
Comment (shreyasseshadri) [reply utils/bst/bst.cpp:72]: As of now I am thinking of supporting delete from a dirty bit, there is no de-allocation of a node while deletion

File: utils/bst/bst.cpp (hunk 3 above)
Comment (srinskit) [reply utils/bst/bst.cpp:93]: Okay. (FYI that's not what a dirty bit is intended for)

**Pull request merged**